### PR TITLE
[ALMIOPEN-46] Batchnorm infer/train/bwds sample apps

### PR DESCRIPTION
**Pull request by SamuelReeder**
## Proposed changes

Provides 3 new hipDNN samples:

1. Batchnorm inference (fully functional)
2. Batchnorm training (functional, but no engine configs)
3. Batchnorm backwards (functional, but no engine configs)

Each of the samples have their own binary, and an explanation in the README.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added automated tests relevant to the introduced functionality
- [x] I have sufficient test coverage for the changes, and code coverage hasn't decreased as a result of my PR
- [x] I have ran the tests, and they are all passing locally
- [x] I have added relevant documentation for the changes
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] I have ran `make format` & `make check_format` to ensure the changes have been formatted

### CODE REVIEW

File: samples/README.MD [hunk 1]
```diff
@@ -1,9 +1,42 @@
...
+        ```
+        where `y` would then be propogated as input to the subsequent layer.
+*   **`bn_training`** (`batchnorm/bn_training.cpp`): Executes the forward pass of a batch normalization training graph on a 4D input tensor.
+    *   For an input `x` of shape `(N, C, H, W)`, the mean and variance are calculated over the `N`, `H`, and `W` dimensions for each of the `C` channels or mini-batches, resulting in a `batch_mean` and `batch_variance` of shape `(1, C, 1, 1)`. It then transforms the input and updates the running statistics:
+        ```
+        y = gamma * (x - batch_mean) / sqrt(batch_variance + eps) + beta
+        running_mean = (1 - momentum) * running_mean + momentum * batch_mean
+        running_variance = (1 - momentum) * running_variance + momentum * batch_variance
+        ```
+    *   The graph outputs the normalized tensor `y`, along with the mini-batch statistics (`saved_mean`, `saved_inv_variance`) required for the backward pass, and the updated population statistics (`running_mean`, `running_variance`) required for inference.
```
Comment (SamuelReeder) [inline samples/README.MD]: [The schema](https://github.com/ROCm/hipDNN/blob/develop/sdk/schemas/batchnorm_attributes.fbs#L17) says its inverse variance, but is it actually referring to inverse standard deviation?

File: samples/batchnorm/bn_backward.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    auto saved_inv_var_tensor
+        = Tensor::make_nchw_tensor<IntermediateType>(saved_inv_variance->get_dim());
+
+    auto dx_tensor = Tensor::make_nchw_tensor<InputType>(dx->get_dim());
+    auto dgamma_tensor = Tensor::make_nchw_tensor<IntermediateType>(dgamma->get_dim());
+    auto dbeta_tensor = Tensor::make_nchw_tensor<IntermediateType>(dbeta->get_dim());
+
+    dy_tensor.template fill_with_random_values<InputType>(static_cast<InputType>(0.0f),
+                                                          static_cast<InputType>(1.0f));
+    dy_tensor.memory().mark_host_modified();
```
Comment (mousdahl-amd) [inline samples/batchnorm/bn_backward.cpp]: You shouldn't need mark_host_modified().  If the device memory's never been accessed, it's assumed that it needs to allocation and copy from the host.

File: samples/CMakeLists.txt [hunk 3]
```diff
@@ -10,14 +10,35 @@ find_package(hipdnn_frontend CONFIG REQUIRED)
...
+add_executable(bn_backward
+    batchnorm/bn_backward.cpp
+)
+target_link_libraries(bn_backward PRIVATE 
+    hip::host
+    hip::device
+    hipdnn_frontend
+    hipdnn_backend
+    hipdnn_sdk
+)
```
Comment (mousdahl-amd) [inline samples/CMakeLists.txt]: If we're going to have a bunch of these, it might make sense to break this into a function.  If we're just going to have a few, probably not worth the effort.

Comment (SamuelReeder) [reply samples/CMakeLists.txt]: That's a good point, I'll do that.

File: frontend/include/hipdnn_frontend/utilities.hpp [hunk 4]
```diff
@@ -70,5 +72,5 @@ inline int32_t initialize_frontend_logging(hipdnnCallback_t fn)
 
     return 0;
 }
-}
```
Comment (BrianHarrisonAMD) [inline frontend/include/hipdnn_frontend/utilities.hpp:88]: Why'd you move this out of the namespace?

File: samples/CMakeLists.txt [hunk 5]
```diff
@@ -2,22 +2,25 @@
 # SPDX-License-Identifier:  MIT
 
 cmake_minimum_required(VERSION 3.25.2)
-project(hipdnn_samples VERSION 0.0.1 LANGUAGES CXX)
+project(hipdnn_samples VERSION 0.1.0 LANGUAGES CXX)
```
Comment (BrianHarrisonAMD) [inline samples/CMakeLists.txt:5]: hehe

File: frontend/include/hipdnn_frontend/utilities.hpp (hunk 4 above)
Comment (SamuelReeder) [reply frontend/include/hipdnn_frontend/utilities.hpp:88]: I thought `hipdnn_frontend::` more suitable than `hipdnn_frontend::graph::`.

File: samples/batchnorm/bn_backward.cpp [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    auto x = create_tensor({16, 16, 16, 16}, input_type);
+    x->set_uid(uid++);
+
+    auto gamma = create_tensor({1, 16, 1, 1}, intermediate_type);
+    gamma->set_uid(uid++);
+
+    auto saved_mean = create_tensor({1, 16, 1, 1}, intermediate_type);
+    saved_mean->set_uid(uid++);
+
+    auto saved_inv_variance = create_tensor({1, 16, 1, 1}, intermediate_type);
```
Comment (BrianHarrisonAMD) [inline samples/batchnorm/bn_backward.cpp]: Might be nice to replace these with NHWC variables to help show how they relate for readability, but fine as is as well.

File: samples/batchnorm/bn_backward.cpp [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    saved_inv_variance->set_uid(uid++);
+
+    auto bn_bwd_attributes = graph::Batchnorm_backward_attributes();
+    bn_bwd_attributes.set_saved_mean_and_inv_variance(saved_mean, saved_inv_variance);
+
+    auto [dx, dgamma, dbeta] = graph->batchnorm_backward(dy, x, gamma, bn_bwd_attributes);
+
+    dx->set_output(true).set_uid(uid++);
+    dgamma->set_output(true).set_uid(uid++);
+    dbeta->set_output(true).set_uid(uid++);
```
Comment (BrianHarrisonAMD) [inline samples/batchnorm/bn_backward.cpp]: You should be able to skip set_uid, and probably worth removing from the samples.
It will automagically manage uid's for you.

File: samples/utils/helpers.hpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            exit(EXIT_FAILURE);                                                                \
+        }                                                                                      \
+    } while(0)
+
+#define HIPDNN_CHECK(status)                                                             \
+    do                                                                                   \
+    {                                                                                    \
+        if(status != HIPDNN_STATUS_SUCCESS)                                              \
+        {                                                                                \
+            std::cerr << "MIOpen Error: " << hipdnnGetErrorString(status) << " in file " \
```
Comment (SamuelReeder) [inline samples/utils/helpers.hpp]: Oops forgot to change this.

File: samples/batchnorm/bn_backward.cpp [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    auto dy_tensor = Tensor::make_nchw_tensor<InputType>(dy->get_dim());
+    auto x_tensor = Tensor::make_nchw_tensor<InputType>(x->get_dim());
+    auto gamma_tensor = Tensor::make_nchw_tensor<IntermediateType>(gamma->get_dim());
+    auto saved_mean_tensor = Tensor::make_nchw_tensor<IntermediateType>(saved_mean->get_dim());
+    auto saved_inv_var_tensor
+        = Tensor::make_nchw_tensor<IntermediateType>(saved_inv_variance->get_dim());
+
+    auto dx_tensor = Tensor::make_nchw_tensor<InputType>(dx->get_dim());
+    auto dgamma_tensor = Tensor::make_nchw_tensor<IntermediateType>(dgamma->get_dim());
+    auto dbeta_tensor = Tensor::make_nchw_tensor<IntermediateType>(dbeta->get_dim());
```
Comment (BrianHarrisonAMD) [inline samples/batchnorm/bn_backward.cpp]: Probably fine as is, but could also add a template option for swapping tensor layouts since it should support either.
If we don't want to add that change, then could leave a comment, or note in the readme.

File: samples/batchnorm/bn_backward.cpp (hunk 6 above)
Comment (SamuelReeder) [reply samples/batchnorm/bn_backward.cpp]: Oh, I didn't know NHWC would work.

File: samples/batchnorm/bn_backward.cpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    std::unordered_map<int64_t, void*> variant_pack;
+    variant_pack[dy->get_uid()] = dy_tensor.memory().template device_data<void>();
+    variant_pack[x->get_uid()] = x_tensor.memory().template device_data<void>();
+    variant_pack[gamma->get_uid()] = gamma_tensor.memory().template device_data<void>();
+    variant_pack[saved_mean->get_uid()] = saved_mean_tensor.memory().template device_data<void>();
+    variant_pack[saved_inv_variance->get_uid()]
+        = saved_inv_var_tensor.memory().template device_data<void>();
+    variant_pack[dx->get_uid()] = dx_tensor.memory().template device_data<void>();
+    variant_pack[dgamma->get_uid()] = dgamma_tensor.memory().template device_data<void>();
+    variant_pack[dbeta->get_uid()] = dbeta_tensor.memory().template device_data<void>();
```
Comment (BrianHarrisonAMD) [inline samples/batchnorm/bn_backward.cpp]: I think we have a clean-up task for this already, but it would be nice to have a better way to grab these buffer pointers. If it exists, can you add a note to clean-up the samples as well?

File: samples/batchnorm/bn_backward.cpp (hunk 9 above)
Comment (SamuelReeder) [reply samples/batchnorm/bn_backward.cpp]: I could do all this with my NHWC ticket, since I didn't think any NHWC tensors would work yet.

File: samples/batchnorm/bn_backward.cpp [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    dx_tensor.memory().mark_device_modified();
+    auto dx_host_ptr = dx_tensor.memory().template host_data<InputType>();
+    std::cout << "First 10 dx values: ";
+    for(int i = 0; i < 10; ++i)
+    {
+        std::cout << static_cast<float>(dx_host_ptr[i]) << " ";
+    }
+
+    std::cout << "\nBatch normalization backward graph execution complete for " << input_type
+              << ".\n\n";
```
Comment (BrianHarrisonAMD) [inline samples/batchnorm/bn_backward.cpp]: Would be nice to add an optional verification step as well just to demonstrate it.
(Might have to leave just stubs in for the implementations that we are adding shortly).

**Review by BrianHarrisonAMD — APPROVED** [review]
Looks good to me!

Some small suggestions on ways to improve, but I think this also is a good start as is.

File: samples/batchnorm/bn_backward.cpp (hunk 6 above)
Comment (BrianHarrisonAMD) [reply samples/batchnorm/bn_backward.cpp]: Whoops sorry I mean replace these dims with NCHW, and specify the values above.
That way people can easily see that the C value is what is used for the 1D tensors.

Like:
```
int64_t N = 16;
int64_t C = 16;
int64_t H = 16;
int64_t W = 16;

auto x = create_tensor{{N, C, H, W}, input_type);
...
```

File: samples/batchnorm/bn_backward.cpp (hunk 9 above)
Comment (SamuelReeder) [reply samples/batchnorm/bn_backward.cpp]: I mean, wouldn't hurt to template it now though!

File: samples/README.MD (hunk 1 above)
Comment (BrianHarrisonAMD) [reply samples/README.MD]: Should be matching with BN training from MIOpen:

https://github.com/ROCm/MIOpen/blob/develop_deprecated/include/miopen/miopen.h#L2733-L2788

File: samples/batchnorm/bn_backward.cpp (hunk 6 above)
Comment (SamuelReeder) [reply samples/batchnorm/bn_backward.cpp]: Ah yes, my (16, 16, 16, 16) tensors with no explanations are definitely a little unclear.

File: samples/README.MD (hunk 1 above)
Comment (BrianHarrisonAMD) [reply samples/README.MD]: But yea it is actually [inverse std](https://github.com/ROCm/MIOpen/blob/develop_deprecated/src/kernels/MIOpenBatchNormActivFwdTrainSpatial.cl#L143).

File: samples/batchnorm/bn_backward.cpp (hunk 10 above)
Comment (mousdahl-amd) [reply samples/batchnorm/bn_backward.cpp]: I've got a note in my Migratable_memory story to do that.  Well, at the very least, try to make it so we don't need the weird template syntax.  That creeps me out.

File: samples/README.MD (hunk 1 above)
Comment (SamuelReeder) [reply samples/README.MD]: Ah thanks, AI said it was called inverse variance purely by convention, but a bit weird still.

**Review by mousdahl-amd — APPROVED** [review]
